File: server/services/postmarkService.ts
import * as postmark from 'postmark';

export interface EmailOptions {
  to: string;
  subject: string;
  htmlContent: string;
  textContent?: string;
  from?: string;
  fromName?: string;
  templateId?: string;
  templateData?: Record<string, any>;
  tag?: string;
  trackOpens?: boolean;
  trackLinks?: 'None' | 'HtmlAndText' | 'HtmlOnly' | 'TextOnly';
}

export interface EmailResult {
  messageId: string;
  status: 'sent' | 'failed';
  error?: string;
  to: string;
  submittedAt: Date;
}

export interface BulkEmailResult {
  successful: number;
  failed: number;
  total: number;
  details: EmailResult[];
}

class PostmarkService {
  private client: postmark.ServerClient | null = null;
  private apiKey: string | undefined;
  private defaultFromEmail: string;
  private defaultFromName: string;

  constructor() {
    this.apiKey = process.env.POSTMARK_API_KEY;
    this.defaultFromEmail = process.env.POSTMARK_FROM_EMAIL || 'noreply@leadgen.ai';
    this.defaultFromName = process.env.POSTMARK_FROM_NAME || 'AI Lead Generation Platform';

    if (this.apiKey) {
      this.client = new postmark.ServerClient(this.apiKey);
      console.log('✅ Postmark email service initialized');
    } else {
      console.warn('⚠️  POSTMARK_API_KEY not configured - email functionality disabled');
    }
  }

  async sendEmail(options: EmailOptions): Promise<EmailResult> {
    if (!this.client) {
      console.warn('Postmark not configured - email not sent');
      return this.getMockEmailResult(options);
    }

    try {
      const emailData: postmark.Models.Message = {
        From: `${options.fromName || this.defaultFromName} <${options.from || this.defaultFromEmail}>`,
        To: options.to,
        Subject: options.subject,
        HtmlBody: options.htmlContent,
        TextBody: options.textContent || this.htmlToText(options.htmlContent),
        Tag: options.tag || 'leadgen-campaign',
        TrackOpens: options.trackOpens !== false,
      };

      // Use template if provided
      if (options.templateId && options.templateData) {
        const templateData = {
          From: emailData.From,
          To: emailData.To,
          Subject: emailData.Subject,
          TemplateId: parseInt(options.templateId),
          TemplateModel: options.templateData,
          Tag: emailData.Tag,
          TrackOpens: emailData.TrackOpens,
          TrackLinks: emailData.TrackLinks,
        };

        const response = await this.client.sendEmailWithTemplate(templateData);
        
        return {
          messageId: response.MessageID,
          status: 'sent',
          to: options.to,
          submittedAt: new Date(response.SubmittedAt),
        };
      } else {
        const response = await this.client.sendEmail(emailData);
        
        return {
          messageId: response.MessageID,
          status: 'sent',
          to: options.to,
          submittedAt: new Date(response.SubmittedAt),
        };
      }
    } catch (error: any) {
      console.error('Postmark email error:', {
        message: error.message,
        errorCode: error.code,
        statusCode: error.statusCode,
        to: options.to,
      });

      return {
        messageId: '',
        status: 'failed',
        error: error.message,
        to: options.to,
        submittedAt: new Date(),
      };
    }
  }

  async sendBulkEmails(emails: EmailOptions[]): Promise<BulkEmailResult> {
    if (!this.client) {
      console.warn('Postmark not configured - bulk emails not sent');
      const details = emails.map(email => this.getMockEmailResult(email));
      return {
        successful: details.length,
        failed: 0,
        total: emails.length,
        details,
      };
    }

    const results: EmailResult[] = [];
    let successful = 0;
    let failed = 0;

    // Postmark supports batch sending up to 500 emails
    const batchSize = 500;
    
    for (let i = 0; i < emails.length; i += batchSize) {
      const batch = emails.slice(i, i + batchSize);
      
      try {
        const emailBatch = batch.map(options => ({
          From: `${options.fromName || this.defaultFromName} <${options.from || this.defaultFromEmail}>`,
          To: options.to,
          Subject: options.subject,
          HtmlBody: options.htmlContent,
          TextBody: options.textContent || this.htmlToText(options.htmlContent),
          Tag: options.tag || 'leadgen-campaign',
          TrackOpens: options.trackOpens !== false,
        }));

        const responses = await this.client.sendEmailBatch(emailBatch);
        
        responses.forEach((response, index) => {
          const originalEmail = batch[index];
          if (response.ErrorCode === 0) {
            successful++;
            results.push({
              messageId: response.MessageID,
              status: 'sent',
              to: originalEmail.to,
              submittedAt: new Date(response.SubmittedAt),
            });
          } else {
            failed++;
            results.push({
              messageId: '',
              status: 'failed',
              error: response.Message,
              to: originalEmail.to,
              submittedAt: new Date(),
            });
          }
        });
      } catch (error: any) {
        console.error('Postmark batch email error:', error);
        
        // Mark all emails in this batch as failed
        batch.forEach(email => {
          failed++;
          results.push({
            messageId: '',
            status: 'failed',
            error: error.message,
            to: email.to,
            submittedAt: new Date(),
          });
        });
      }
    }

    return {
      successful,
      failed,
      total: emails.length,
      details: results,
    };
  }

  async getDeliveryStats(messageId: string): Promise<any> {
    if (!this.client) {
      return null;
    }

    try {
      const outboundMessage = await this.client.getOutboundMessageDetails(messageId);
      return {
        messageId,
        status: outboundMessage.Status,
        events: outboundMessage.MessageEvents,
        receivedAt: outboundMessage.ReceivedAt,
        subject: outboundMessage.Subject,
        to: outboundMessage.Recipients,
      };
    } catch (error) {
      console.error('Error getting delivery stats:', error);
      return null;
    }
  }

  async getAccountInfo(): Promise<any> {
    if (!this.client) {
      return null;
    }

    try {
      // Note: Account info endpoint may require specific API permissions
      // For now, return basic service status
      return {
        service: 'Postmark',
        configured: true,
        fromEmail: this.defaultFromEmail,
        fromName: this.defaultFromName,
      };
    } catch (error) {
      console.error('Error getting account info:', error);
      return null;
    }
  }

  private htmlToText(html: string): string {
    // Basic HTML to text conversion
    return html
      .replace(/<br\s*\/?>/gi, '\n')
      .replace(/<\/p>/gi, '\n\n')
      .replace(/<[^>]*>/g, '')
      .replace(/&nbsp;/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'")
      .trim();
  }

  private getMockEmailResult(options: EmailOptions): EmailResult {
    return {
      messageId: `mock_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      status: 'sent',
      to: options.to,
      submittedAt: new Date(),
    };
  }

  isConfigured(): boolean {
    return !!this.client;
  }

  getConfiguration(): { configured: boolean; fromEmail?: string; fromName?: string } {
    return {
      configured: this.isConfigured(),
      fromEmail: this.defaultFromEmail,
      fromName: this.defaultFromName,
    };
  }
}

export const postmarkService = new PostmarkService();

File: backend/services/postmark_service.py
"""
Postmark email service integration
"""
import os
from postmark import PMMail
import logging
from typing import Optional

logger = logging.getLogger(__name__)

class PostmarkService:
    def __init__(self):
        self.api_key = os.getenv("POSTMARK_API_KEY")
        if not self.api_key:
            logger.warning("POSTMARK_API_KEY not found - email functionality disabled")
    
    def send_email(
        self,
        to: str,
        subject: str,
        content: str,
        from_email: str = "noreply@leadgen.ai",
        from_name: str = "AI Lead Generation Platform"
    ) -> bool:
        """Send email via Postmark"""
        if not self.api_key:
            logger.error("Postmark API key not configured")
            return False
        
        try:
            message = PMMail(
                api_key=self.api_key,
                subject=subject,
                sender=f"{from_name} <{from_email}>",
                to=to,
                html_body=content,
                text_body=self._html_to_text(content)
            )
            
            result = message.send()
            
            if result.get('ErrorCode') == 0:
                logger.info(f"Email sent successfully to {to}")
                return True
            else:
                logger.error(f"Email failed to {to}: {result.get('Message')}")
                return False
                
        except Exception as e:
            logger.error(f"Error sending email to {to}: {e}")
            return False
    
    def send_bulk_emails(self, emails: list) -> dict:
        """Send multiple emails in batch"""
        if not self.api_key:
            logger.error("Postmark API key not configured")
            return {"success": 0, "failed": len(emails)}
        
        success_count = 0
        failed_count = 0
        
        for email_data in emails:
            success = self.send_email(
                to=email_data["to"],
                subject=email_data["subject"],
                content=email_data["content"],
                from_email=email_data.get("from_email", "noreply@leadgen.ai"),
                from_name=email_data.get("from_name", "AI Lead Generation Platform")
            )
            
            if success:
                success_count += 1
            else:
                failed_count += 1
        
        return {
            "success": success_count,
            "failed": failed_count,
            "total": len(emails)
        }
    
    def _html_to_text(self, html_content: str) -> str:
        """Convert HTML content to plain text"""
        # Simple HTML to text conversion
        import re
        
        # Remove HTML tags
        text = re.sub('<[^<]+?>', '', html_content)
        
        # Clean up whitespace
        text = re.sub(r'\s+', ' ', text).strip()
        
        return text

# Global Postmark service instance
postmark_service = PostmarkService()